### PR TITLE
[SEMI-MODULAR] [TEST MERGE ONLY] General baton rework 2: Electric Boogaloo, batons-don't-suck edition (Now with OVERCLOCKING)

### PR DESCRIPTION
**Pull request by nikothedude**
# _TEST-MERGE FIRST_

## About The Pull Request

FAQ: https://hackmd.io/@NikoTheGuyDUde/HJvHskdI3

Branched off https://github.com/Skyrat-SS13/Skyrat-tg/pull/20834 as I searched for an alternate solution. Expanding apon junglerat's idea of an alternate mode, I created it, tested it, and realized, "God damn this is cool and I should really share this."
So here it is!

Reworks batons to respect melee armor and not be omnipresent on the game's balance scale.

Normal batons knockdown near-instantly but suffer against armor in terms of stamina damage. This relegates them mostly to self-defense duties, which is fine when they are most commonly found on non-combat roles (Heads of staff, detective)

The contractor baton's damage has been reduced to 85, and it suffers somewhat against heavy armor to allow security to more properly prepare against contractors and encourage them to use more of their kit.

The stun baton, while suffering against armor and having knockdown be on the second hit, now has a alternate mode: **Overcharge**
Overcharge massively increases armor penetration and moves knockdown back to the first hit, at the cost of extreme stun cell cost and passive discharge. 

See testing proof for videos/screenshots of this PR in action.

Read the design doc for more specific info.

Design doc: https://hackmd.io/@C9920KwxTou_TP6xAfZ8FA/BJrG7cC7h

## How This Contributes To The Skyrat Roleplay Experience

Read the design doc for the rationale.

## Proof of Testing

<details>

image

How to interpret: 
x- = You have enough for EXACTLY this amount of charges. Once you use them both, you have less than the charge youd need for a normal stun.
x.# = Higher decimal = higher leftover charge after all overcharges are used.

image

The small glow
image

https://user-images.githubusercontent.com/59709059/235787913-a3a00785-f3eb-4e19-914c-f6799212d24c.mp4

https://user-images.githubusercontent.com/59709059/235786498-ca76a81e-19e6-4c82-b633-df590f695be0.mp4

https://user-images.githubusercontent.com/59709059/235786548-816768da-01fd-4bbc-be1b-a6562e8ae8df.mp4

<summary>Screenshots/Videos</summary>
  
</details>

## Changelog

:cl:
balance: Batons now respect armor
balance: Police/Telebatons are now relegated to self-defense duties, having near-guaranteed knockdown but with very lacking performance against armor
balance: Stun batons now induce knockdown on second hit
balance: Contractor baton stamina damage (105 -> 85)
balance: Batons now have a small delay for inserting a new cell
add: New baton mode: OVERCHARGE. Massively increase stun cell cost and passive discharge in return for extreme armor penetration and easier knockdown.
/:cl:

### CODE REVIEW

**Comment by nikothedude** [comment]
I'm not sure if this has a chance, but I think it's damn cool and promising enough to at least let people look at it

**Comment by nikothedude** [comment]
Also, this will only be undrafted once consensus is reached

**Comment by nikothedude** [comment]
update: ive talked enough to know people are at least willing to give it a chance so im opening it in case anyone wants to test merge it

also if yo utest merge it fuckin ask me first so i can watch the round

**Comment by tf-4** [comment]
ask someone to reopen who's willing to TM

**Comment by nikothedude** [comment]
i should mention im not working on my prs rn becuase im hyperfocusing on modding starsector, once thats done ill return

**Comment by nikothedude** [comment]
oh also im ready again wo

**Comment by KathrinBailey** [comment]
Keep as draft until it's ready for review.

**Comment by YakumoChen** [comment]
it would be nice if the overcharge mode had a way to use it besides the action button, which isn't really good for combat.

**Comment by Eastwardrope** [comment]
So some comments on this from my personal experiences. Contractors already suffer greatly against security early game, their baton isn't good against the 12 plus people who are on their ass due to the fact 1 it can't lethally hit anyone 2 it has a fairly long cooldown(compared to normal batons which hit faster on the stun mode normally). Both of these effects mean the contractor can't effectively fight security given the fact they are always coming at him in swarms making the baton useless against heavy armor makes that baton functionally useless against security and early game that maybe the only weapon they have period. 2 shotting someone with that baton is it's whole thing due to the fact it has such a long cooldown(It also occassionally won't force people to drop their weapons resulting in the contractor getting blasted because he can't stun multiple people effectively)

As for the overcharge I think it will result in only ever being used unless it completely drains the battery of the baton in that one hit. I think as well you'd need to remove the second baton all security officers get to make that work properly because I can very well see an officer overcharge batoning someone so they are forced to crawl then quick drawing the second baton and using it for the rest negating any negatives the overcharge did. Or making the overcharge deal massive stamina damage to the officer or otherwise be a negative so it isn't just the opener in every fight always.

Additionally this is a question about the telebaton/police baton part. Does the knockdown disarm? Because if thats the case it seems more potent than the normal baton for an opener if it knockdowns and disarms on 1 hit through armor.

Also what happened to stun prods through this? Are they the same?

**Comment by nikothedude** [comment]
stun prods are the same, they have a 2x cost mult on all swings already so i think theyre fine

i cant just remove the baton, i have to make using multiple at once inefficient, which i can achieve by cranking up the passive discharge to 11 if need be

normal batons have always been way better for knockdown, this didnt change that - they still instaknock

also, contractor being bad against sec in the earlygame with 12+ people running at them honestly isnt much of a concern to me, contractors arent really meant to be able to stand up to the entire secteam regardless
maybe ill come up with some solution specific to contractor to make them more vialb,e but i'd need to be more convinced that contractor NEEDS to be able to stand up against security so well

**Comment by nikothedude** [comment]
> it would be nice if the overcharge mode had a way to use it besides the action button, which isn't really good for combat.

you can actually clickdrag the button around for more convenient use

**Comment by nikothedude** [comment]
update will come tomorrow to address some issuessssss

**Comment by Eastwardrope** [comment]
> 

Contractors don't need to fight all of sec but making the baton unable to deal with armor makes dealing with sec early on impossible irregardless of how many of them there. Being able to deal with an officer in any gear is the batons strong suit as some of your targets are sec and have to be dealt with somehow. Making their baton unable to deal with even a normal officer means they 100 percent can not do all their contracts that they have. If you wanna nerf Contractors nerf their armor since it can easily tank a shotgun to the chest while being mobile.

**Comment by anconfuzedrock** [comment]
Moving the button is still really inconvenient, I'd rather a hotkey. alas, TG isn't a macro server.
Main issue I see is that with a regular battery, overcharge takes a second to charge and actually only lasts like 2 seconds. I'm not sure what balance it can solve when it's _very_ likely a sec player will just run out of charge if they try to use it (or on a big nasty like carp, probably not even have time to charge it). but for now it's just really inconvenient with a standard battery. could change it from needing 95% charge to 90% charge.

**Comment by nikothedude** [comment]
patch notes https://hackmd.io/cl8uH4BkSCirBWvPi-_gLw

**Comment by nikothedude** [comment]
starting to compile questions into a FAQ, ctrl + f the topic you want to see https://hackmd.io/@NikoTheGuyDUde/HJvHskdI3

**Comment by anconfuzedrock** [comment]
the batteries being unusable on removal just seems kinda mean. see: any accident where the battery gets removed too early. Prank your local sec officer by disarming them and screwdrivering the baton. Is it that important?
edit: and with the recharge being basically non-existent it is- I'd genuinely rather just have a cell with a slightly bigger battery than highcap than this thing.

**Comment by YakumoChen** [comment]
Put supercap cells in the premium category in secvend, give officers an option to spend their round start money on if they can't wait 15 minutes for science to make bluespace cells

The unrechargable cells are a little silly since security has virtually nothing else to use power cells for but whatever works I guess. There's hardly a reason to even use Batons early in the round except on stupid assistants or to hit carp (which do not take stamina damage)

**Comment by YakumoChen** [comment]
Also the overcharge should have an EMP effect (like electrocuting the officer holding it) because EMP effects are cool.

EMP is already supposed to be an anti-security antag tool and Batons should definitely not be spared from it.

**Comment by nikothedude** [comment]
> the batteries being unusable on removal just seems kinda mean. see: any accident where the battery gets removed too early. Prank your local sec officer by disarming them and screwdrivering the baton. Is it that important? edit: and with the recharge being basically non-existent it is- I'd genuinely rather just have a cell with a slightly bigger battery than highcap than this thing.

good point on the removal shit being mean theres an alternate method to doing this

**Comment by YakumoChen** [comment]
How does this effect dept guard batons

**Comment by nikothedude** [comment]
> How does this effect dept guard batons

same as normal batons, i repeat what i say: just keep hitting them and they go down eventually with a good cell

**Comment by YakumoChen** [comment]
Guard Batons don't rely on power cells...?

**Comment by nikothedude** [comment]
> Guard Batons don't rely on power cells...?

They do.

**Comment by YakumoChen** [comment]
They don't, they explicitly rely on "uses" instead of power cell charge

**Comment by YakumoChen** [comment]
Have you checked if guard Batons have overcharge because they probably shouldnt

**Comment by nikothedude** [comment]
> They don't, they explicitly rely on "uses" instead of power cell charge

They do. Look at your baton next time you use them and youll see they use charge

**Comment by nikothedude** [comment]
> Have you checked if guard Batons have overcharge because they probably shouldnt

They do and they should

**Comment by nikothedude** [comment]
https://hackmd.io/@NikoTheGuyDUde/SkrH4QO82 more patch notes

**Comment by YakumoChen** [comment]
why should guard batons have overcharge? they're meant for defending departments, not validhunting heavily armoured antags.

**Comment by nikothedude** [comment]
> why should guard batons have overcharge? they're meant for defending departments, not validhunting heavily armoured antags.

For cases where heavily armored antags are attacking the department

**Comment by YakumoChen** [comment]
> > why should guard batons have overcharge? they're meant for defending departments, not validhunting heavily armoured antags.
> 
> For cases where heavily armored antags are attacking the department

That's what security is for, not guards

Guards are intentionally given lower-powered security gear because they are not a replacement for security

**Comment by nikothedude** [comment]
new update, shift+ f is now the hotkey for overcharging

should help with the clunkiness

**Comment by projectkepler-ru** [comment]
I loved the way baton worked with direct stamina damage as opposed to knock down. 
but I think this PR is a little silly and it's kind of a pain to use

**Comment by Paxilmaniac** [comment]
> new update, shift+ f is now the hotkey for overcharging
> 
> should help with the clunkiness

I had a suggestion upon seeing this logging in today,

image

That being, if there's going to be a hotkey setup for setting batons to overcharge, don't have it be its own completely unique keybind that only batons use. I'd suggest something like overcharge turning on if you right click either the baton or just anything that isn't a mob while holding the baton.

Having a unique keybind that only batons use and that literally nothing else does is 1. Not an ideal use of keybinds I think, 2. Not very good interface for someone who's either never used a baton before or did not take the time to read through the entire pr + its comments and whatnot.

**Comment by nikothedude** [comment]
> > new update, shift+ f is now the hotkey for overcharging
> > should help with the clunkiness
> 
> I had a suggestion upon seeing this logging in today,
> 
> image
> 
> That being, if there's going to be a hotkey setup for setting batons to overcharge, don't have it be its own completely unique keybind that only batons use. I'd suggest something like overcharge turning on if you right click either the baton or just anything that isn't a mob while holding the baton.
> 
> Having a unique keybind that only batons use and that literally nothing else does is 1. Not an ideal use of keybinds I think, 2. Not very good interface for someone who's either never used a baton before or did not take the time to read through the entire pr + its comments and whatnot.

The specific suggestion of right click doesnt work - This removes their ability to bash things
Middle click doesnt work either - Already used for swapping hands
Modifier keys + mouse keys also dont work, as the code doesnt support this

Generifying the hotkey to be used with something else would be good, yes
Actually I was going to counterargue but then I realized I could maybe move it to shift+Z as this is the most logically intuitive place to put the hotkey, as it's a "mode switch"

**Comment by nikothedude** [comment]
_Maybe_ I could move it to shift+z and generify it so that it's instead a generic "alternate mode switch of held item", but for now, I think it's _fine_, especially if I just document the hotkey a bit more in-game (It's already on the action button)

**Comment by Paxilmaniac** [comment]
> 

I didn't say just right click, its right click _if you aren't hitting anything_. For a function as fundamental to the baton's use as overcharge I really, really suggest you make it some method of activating that's comparable to like literally anything else in the game. There is a reason we don't do special hotkeys for anything ever relating to items, just try and imagine playing the game but every single item with some special action had its own hotkey you have to set. Now imagine if those hotkeys are _unbound by default_, meaning you couldn't even use any of those things without going into your settings.

image

**Comment by nikothedude** [comment]
> > 
> 
> I didn't say just right click, its right click _if you aren't hitting anything_. For a function as fundamental to the baton's use as overcharge I really, really suggest you make it some method of activating that's comparable to like literally anything else in the game. There is a reason we don't do special hotkeys for anything ever relating to items, just try and imagine playing the game but every single item with some special action had its own hotkey you have to set. Now imagine if those hotkeys are _unbound by default_, meaning you couldn't even use any of those things without going into your settings.
> 
> image

Yeah I agree the whole snowflaked hotkey thing is and was off to me, Im just onl doing it because other options seemed more janky 

Ill see about moving it to a more generic hotkey/whatever later when I'm not exhausted

**Comment by YakumoChen** [comment]
Right click while hitting the floor wouldn't be very ideal since it interferes with the default baton actions. You'd be overcharging while trying to clickspam harm baton mobs in combat mode.

File: modular_skyrat/modules/sec_haul/code/melee/baton.dm [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+ * Often is called after a delay from toggle_overcharge.
+ * This delay, coupled with the passive discharge, encourages users to only keep overcharge active
+ * when absolutely necessary, and think stratetegically about when to use it.
+ *
+ * Plus, the high cell cost and discharge both incentivise users to upgrade their cell, which is always good.
+ *
+ * Args:
+ * mob/user: The person that initiated the overcharge attempt. Nullable.
+ */
+/obj/item/melee/baton/security/proc/enable_overcharge(mob/user)
```
Comment (YakumoChen) [inline modular_skyrat/modules/sec_haul/code/melee/baton.dm:259]: You must add a chunky fingers check so overcharge cannot be enabled while under a condition the stun cannot be used normally.

**Comment by nikothedude** [comment]
also on the backburner for a bit im not sure why it still says its test merged

File: modular_skyrat/modules/sec_haul/code/melee/baton.dm [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+/// The volume of the extra sound played on overcharged baton impact, for extra "oomph" factor
+#define BATON_OVERCHARGE_EXTRA_HITSOUND_VOLUME 20
+
+/// How long it takes for a user to replace a cell in a baton.
+#define BATON_CELL_REPLACE_TIME 3 SECONDS
+
+// Police baton - very hard to not get knocked down by it, though its low attack rate limits its effectiveness
+/obj/item/melee/baton
+	/// The armor flag used when we use our stun function, AKA our left click.
+	var/stun_armor_flag = MELEE
```
Comment (KathrinBailey) [inline modular_skyrat/modules/sec_haul/code/melee/baton.dm:19]: Needs to be set to energy for stunbatons and kept melee for regular truncheon-like batons like the telescopic baton.

**Pull request closed without merging**